src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py

```diff
@@ -0,0 +1,337 @@
+
+    # from the original Blip Diffusion code, speciefies the target subject and augments the prompt by repeating it
+    def _build_prompt(self, prompts, tgt_subjects, prompt_strength=1.0, prompt_reps=20):
+        rv = []
+        for prompt, tgt_subject in zip(prompts, tgt_subjects):
+            prompt = f"a {tgt_subject} {prompt.strip()}"
+            # a trick to amplify the prompt
+            rv.append(", ".join([prompt] * int(prompt_strength * prompt_reps)))
+
+        return rv
+
+    def prepare_latents(self, batch_size, num_channels, height, width, dtype, device, generator, latents=None):
+        shape = (batch_size, num_channels, height, width)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
+            )
+
+        if latents is None:
+            latents = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            latents = latents.to(device=device, dtype=dtype)
+
+        return latents
+
+    def encode_prompt(self, query_embeds, prompt):
+        #embeddings for prompt, with query_embeds as context
+        tokenized_prompt = self._tokenize_text(prompt).to(self.device)
+        text_embeddings = self.text_encoder(
+            input_ids=tokenized_prompt.input_ids,
+            ctx_embeddings=query_embeds,
+            ctx_begin_pos=[self.config.ctx_begin_pos],
+        )[0]
+
+        return text_embeddings
+
+    def preprocess_caption(self, caption):
+        caption = re.sub(
+            r"([.!\"()*#:;~])",
+            " ",
+            caption.lower(),
+        )
+        caption = re.sub(
+            r"\s{2,}",
+            " ",
+            caption,
+        )
+        caption = caption.rstrip("\n")
+        caption = caption.strip(" ")
+
+        # truncate caption
+        caption_words = caption.split(" ")
+        if len(caption_words) > 50:
+            caption = " ".join(caption_words[:50])
+
+        return caption
+    
+
+    @torch.no_grad()
+    def __call__(
+        self,
+        prompt,
+        reference_image,
+        source_subject_category,
+        target_subject_category,
+        condtioning_image=None,
+        latents=None,
+        guidance_scale=7.5,
+        height=512,
+        width=512,
+        seed=42,
+        num_inference_steps=50,
+        neg_prompt="",
+        prompt_strength=1.0,
+        prompt_reps=20,
+        use_ddim=False,
+    ):
+        prompt = [self.preprocess_caption(prompt)]
+        source_subject_category = [self.preprocess_caption(source_subject_category)]
+        target_subject_category = [self.preprocess_caption(target_subject_category)]
+
+        mean = (0.48145466, 0.4578275, 0.40821073)
+        std = (0.26862954, 0.26130258, 0.27577711)
+
```

Comment:
These two should be configuration variables.

Reply:
Done